cedar-policy-validator/src/types.rs

```diff
@@ -1376,11 +1404,57 @@ impl AttributeType {
         }
     }
 
-    /// Construct an [`AttributeType`] for an attribute that must be present given
-    /// the type of the attribute.
+    /// Construct an [`AttributeType`] for an attribute that is required.
     pub fn required_attribute(attr_type: Type) -> Self {
         Self::new(attr_type, true)
     }
+
+    /// Construct an [`AttributeType`] for an attribute that is optional.
+    pub fn optional_attribute(attr_type: Type) -> Self {
+        Self::new(attr_type, false)
+    }
+
+    /// Is the attribute required?
+    pub fn is_required(&self) -> bool {
+        self.is_required
+    }
+
+    /// Display just the type portion of the [`AttributeType`], ignoring the
+    /// `is_required` flag
+    fn display(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.attr_type.fmt(f)
+    }
+
+    /// Get the least upper bound of two [`AttributeType`]s
+    fn least_upper_bound(
+        schema: &ValidatorSchema,
+        ty0: &AttributeType,
+        ty1: &AttributeType,
+        mode: ValidationMode,
+    ) -> Result<AttributeType, LubHelp> {
+        Type::least_upper_bound(schema, &ty0.attr_type, &ty1.attr_type, mode).and_then(|lub| {
+            let is_lub_required = ty0.is_required() && ty1.is_required();
+            if mode.is_strict() && ty0.is_required() != ty1.is_required() {
+                Err(LubHelp::AttributeQualifier)
+            } else {
+                Ok(AttributeType::new(lub, is_lub_required))
+            }
+        })
+    }
+
+    /// Is `ty0` a subtype of `ty1`?
+    fn is_subtype(
+        schema: &ValidatorSchema,
+        ty0: &AttributeType,
+        ty1: &AttributeType,
+        mode: ValidationMode,
+    ) -> bool {
+        (if mode.is_strict() {
```

Comment:
While we're moving the code around I might also split this up with a local for the result of optional-qualifier subtyping

Reply:
Done